File: Algorithm/maximum_gap.py
class Solution:
    def maximumGap(self,num):
        length=len(num)
        if length<2:
            return 0
        num=self.radixSort(num)
        res=0
        for i in range(1,length):
            res=max(num[i]-num[i-1],res)
        return res
    def radixSort(self,num):
        for k in range(31):
            s=[[] for i in range(2)]
            for i in num:
                s[i/(2**k)%2].append(i)
            num=[a for b in s for a in b]
        return num
    def maximumGap1(self,num):
        length=len(num)
        if length<2:
            return 0
        num = self.radixSort1(num)
        res=0
        for i in range(1,length):
            if res < num[i] - num[i-1]:
                res = num[i] - num[i-1]
        return res
    def radixSort1(self,num):
        for i in range(31):
            bucket =[[] for j in range(2)]
      
            for val in num:
                if (val >> i) & 1:
                    bucket[1].append(val)
                else:
                    bucket[0].append(val)  
            num = []
            num += bucket[0]
            num += bucket[1]
        return num
if __name__=='__main__':
    sol=Solution()
    num=[100,3,2,1]
    print sol.maximumGap(num)
    print sol.maximumGap1(num)

File: Algorithm/CPP/maximum_gap.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <cctype>
# include <algorithm>
# include<time.h>
using namespace std;


class Solution {
public:
	int maximumGap(vector<int>& nums) {
		int n = nums.size();
		if (n < 2)return 0;
		int res; 
		vector<int> num(radixSort(nums));
		for (int i = 1; i < n; i++)
			res = max(res, num[i] -num[i - 1]);
		return res;
	}
	vector<int> radixSort(vector<int>& nums)
	{
		int mask = 0;
		for (int i = 0; i < 32; i++)
		{
			vector<vector<int>>  bucket(2);
			mask = 1 << i;
			for (auto num : nums)
			{
				if (num&mask)
				{
					bucket[1].push_back(num);
				}
				else{
					bucket[0].push_back(num);
				}
			}
			nums.clear();
			for (auto lst : bucket)
				for (auto val : lst)
					nums.push_back(val);
		}
		return nums;
	}
};

int main()
{
	Solution sol;
	vector<int> nums = { 1, 3, 2, 8 };
	cout << sol.maximumGap(nums) << endl;
	for (auto k : sol.radixSort(nums)) cout << k << " ";
	cout << endl;

		
			
	system("pause");
	return 0;
}

